# 1. Functions

In Python, a function is a reusable block of code that performs a specific task or set of tasks. Functions are defined using the `def` keyword, and they can accept input parameters (arguments) and return a result using the `return` statement. Here's the basic syntax for defining a function:

```python
def function_name(parameters):
    # Function code here
    # ...
    return result
```

Here are some examples of Python functions:

1. **Simple Function** - A function that takes no arguments and returns a greeting message.

```python
def greet():
    return "Hello, World!"

# Call the function
message = greet()
print(message)
```

2. **Function with Parameters** - A function that takes two numbers as arguments and returns their sum.

```python
def add(a, b):
    return a + b

# Call the function
result = add(3, 5)
print(result)
```

3. **Function with Default Parameters** - A function that takes an optional parameter with a default value.

```python
def greet(name="User"):
    return f"Hello, {name}!"

# Call the function with and without a name
print(greet())  # Output: Hello, User!
print(greet("Alice"))  # Output: Hello, Alice!
```

4. **Function with Multiple Returns** - A function that returns multiple values using tuple unpacking.

```python
def rectangle_info(length, width):
    area = length * width
    perimeter = 2 * (length + width)
    return area, perimeter

# Call the function and unpack the results
area, perimeter = rectangle_info(4, 6)
print(f"Area: {area}, Perimeter: {perimeter}")
```

5. **Recursive Function** - A function that calls itself to compute the factorial of a number.

```python
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)

result = factorial(5)  # Calculates 5!
print(result)
```

6. **Lambda Function** - A small anonymous function defined using the `lambda` keyword.

```python
square = lambda x: x ** 2
print(square(4))  # Output: 16
```

7. **Function with Variable-Length Arguments** - A function that accepts a variable number of arguments using `*args` and calculates their sum.

```python
def sum_all(*args):
    total = 0
    for num in args:
        total += num
    return total

result = sum_all(1, 2, 3, 4, 5)
print(result)  # Output: 15
```

These are just some basic examples of Python functions. Functions are a fundamental concept in programming and can be used to encapsulate logic, promote code reusability, and improve the organization of your code.

# 2. Print vs Return in Functions

In Python, both `print` and `return` are used in functions, but they serve different purposes.

1. `print` in a Python function:
   - `print` is a built-in function in Python that is used to display information on the console or terminal.
   - When you use `print` within a function, it will output the specified text or variables to the console, but it won't affect the value returned by the function itself.
   - `print` is typically used for debugging or displaying information to the user but doesn't impact the program's logic or data flow.
   - Example:

    ```python
    def say_hello(name):
        print("Hello, " + name)
    
    say_hello("Alice")  # This will print "Hello, Alice" to the console but won't return a value.
    ```

2. `return` in a Python function:
   - `return` is used to specify the value that a function should produce as its result.
   - When a function encounters a `return` statement, it immediately exits the function and returns the specified value to the caller.
   - The value returned by a function can be stored in a variable or used in further computations.
   - Example:

    ```python
    def add(a, b):
        return a + b
    
    result = add(3, 5)  # The result variable will store the value 8.
    ```

In summary, `print` is used for displaying information on the console, while `return` is used to provide a result or value that can be used by the calling code. It's important to differentiate between these two when writing functions in Python, as they serve distinct purposes.

# Practice

In [1]:
def helloworld():
    print('Hello World') 
    
var = helloworld() # With None Type, you cannot perform any primitive operations.
type(var)

Hello World


NoneType

In [2]:
def helloworld():
    return 'Hello World'
    
var = helloworld()
type(var)

str

In [3]:
def helloworld():
    return 'Hello World', 100, 5.5, True, 3+6j
    
var = helloworld()
type(var)

tuple

In [4]:
for i in var:
    print(type(i))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>
<class 'complex'>


In [5]:
def some_function():
    return True, False, False, True

a,b,c,d = some_function()

print(a,b,c,d)

True False False True


In [1]:
database = {
    'siddharth@gmail.com':'pwskills@12345',
    'sudhanshu@gmail.com':'ineuron@12345'
}

def login(email,password):

    if email in database.keys():
        if password == database[email]:
            print('Login Successful.')
        else:
            print('Wrong Password. Login Failed!')
    else:
        print('Email address is not registered.')

In [2]:
login('siddharth1001@gmail.com','random')

Email address is not registered.


In [3]:
login('siddharth@gmail.com','wrongpassword')

Wrong Password. Login Failed!


In [4]:
login('siddharth@gmail.com','pwskills@12345')

Login Successful.


# 3. Type Hinting

In Python, you do not explicitly specify the data type of function parameters. Python is a dynamically typed language, which means that the data type of a variable is determined at runtime, and you do not need to declare the data type of a parameter when defining a function.

Here's an example of a Python function without explicitly specifying data types for its parameters:

```python
def add_numbers(a, b):
    return a + b

result = add_numbers(5, 3)
print(result)  # Output: 8
```

In the `add_numbers` function, `a` and `b` are parameters, and you can pass any type of data to them (e.g., integers, floats, strings). Python will perform type inference based on the actual arguments passed to the function.

However, if you want to add type hints for function parameters to improve code readability and provide hints to developers and tools like linters and type checkers, you can use the "Type Hinting" feature introduced in Python 3.5 and later. Type hints do not enforce data types at runtime but provide information about the expected types:

```python
def add_numbers(a: int, b: int) -> int:
    return a + b

result = add_numbers(5, 3)
print(result)  # Output: 8
```

In the modified example above, we've added type hints using the `:` syntax to indicate that `a` and `b` are expected to be of type `int`, and the `-> int` annotation specifies that the function will return an `int`. These hints are for documentation and can be used by tools like `mypy` for static type checking. They do not enforce types at runtime; Python remains dynamically typed.

In [5]:
def add_numbers(a, b):
    return a + b

result = add_numbers(5, 3)
print(result)  # Output: 8

8


In [6]:
def add_numbers(a: int, b: int) -> int:
    return a + b

In [7]:
help(add_numbers)

Help on function add_numbers in module __main__:

add_numbers(a: int, b: int) -> int



In [8]:
result = add_numbers(5, 5)
print(result)  # Output: 8

10


In [9]:
result = add_numbers('Add','This')
print(result)  # Output: AddThis

AddThis


# 4. Docstring

A docstring in Python is a special type of string that is used to provide documentation for functions, classes, modules, or methods. Docstrings are typically placed at the beginning of these code elements and are enclosed in triple quotes (either single or double). They serve as a way to describe what the code does, provide information about its usage, and offer insights into its purpose.

Here's a basic example of a Python function with a docstring:

```python
def greet(name):
    """
    This function takes a name as an argument and returns a greeting message.
    
    Parameters:
    name (str): The name of the person to greet.
    
    Returns:
    str: A greeting message that includes the name.
    """
    return f"Hello, {name}!"
```

In this example:

- The docstring is enclosed in triple double quotes (`"""`) and is placed immediately below the function definition.
- It provides a brief description of what the function does.
- It documents the parameters the function accepts, their types, and their meanings.
- It specifies the type of the return value and provides a description of what the function returns.

Docstrings are not just comments; they are accessible programmatically and can be accessed using the `help()` function or through tools like Sphinx for generating documentation.

Here's how you can access the docstring of the `greet` function:

```python
help(greet)
```

This will display the docstring, including the description, parameter details, and return value information.

Using meaningful and descriptive docstrings is considered good practice in Python as it makes code more understandable and helps other developers (and your future self) when using or maintaining the code.

In [10]:
def greet(name):
    """
    This function takes a name as an argument and returns a greeting message.
    
    Parameters:
    name (str): The name of the person to greet.
    
    Returns:
    str: A greeting message that includes the name.
    """
    return f"Hello, {name}!"

In [11]:
help(greet)

Help on function greet in module __main__:

greet(name)
    This function takes a name as an argument and returns a greeting message.
    
    Parameters:
    name (str): The name of the person to greet.
    
    Returns:
    str: A greeting message that includes the name.



# 5. Docstring Examples

In [12]:
help(print)

Help on built-in function print in module builtins:

print(*args, sep=' ', end='\n', file=None, flush=False)
    Prints the values to a stream, or to sys.stdout by default.
    
    sep
      string inserted between values, default a space.
    end
      string appended after the last value, default a newline.
    file
      a file-like object (stream); defaults to the current sys.stdout.
    flush
      whether to forcibly flush the stream.



In [13]:
help(type)

Help on class type in module builtins:

class type(object)
 |  type(object) -> the object's type
 |  type(name, bases, dict, **kwds) -> a new type
 |  
 |  Methods defined here:
 |  
 |  __call__(self, /, *args, **kwargs)
 |      Call self as a function.
 |  
 |  __delattr__(self, name, /)
 |      Implement delattr(self, name).
 |  
 |  __dir__(self, /)
 |      Specialized __dir__ implementation for types.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __init__(self, /, *args, **kwargs)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __instancecheck__(self, instance, /)
 |      Check if an object is an instance.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __setattr__(self, name, value, /)
 |      Implement setattr(self, name, value).
 |  
 |  __sizeof__(self, /)
 |      Return mem

In [15]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

In [16]:
help(min)

Help on built-in function min in module builtins:

min(...)
    min(iterable, *[, default=obj, key=func]) -> value
    min(arg1, arg2, *args, *[, key=func]) -> value
    
    With a single iterable argument, return its smallest item. The
    default keyword-only argument specifies an object to return if
    the provided iterable is empty.
    With two or more arguments, return the smallest argument.



In [17]:
help(max)

Help on built-in function max in module builtins:

max(...)
    max(iterable, *[, default=obj, key=func]) -> value
    max(arg1, arg2, *args, *[, key=func]) -> value
    
    With a single iterable argument, return its biggest item. The
    default keyword-only argument specifies an object to return if
    the provided iterable is empty.
    With two or more arguments, return the largest argument.



In [18]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



In [19]:
help(range)

Help on class range in module builtins:

class range(object)
 |  range(stop) -> range object
 |  range(start, stop[, step]) -> range object
 |  
 |  Return an object that produces a sequence of integers from start (inclusive)
 |  to stop (exclusive) by step.  range(i, j) produces i, i+1, i+2, ..., j-1.
 |  start defaults to 0, and stop is omitted!  range(4) produces 0, 1, 2, 3.
 |  These are exactly the valid indices for a list of 4 elements.
 |  When step is given, it specifies the increment (or decrement).
 |  
 |  Methods defined here:
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getitem__(self, key, /)
 |      Return self[key].
 |  
 |  __gt__(self, value, /)
 |      Return self>value.
 |  
 |  __hash

In [20]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [21]:
help(input)

Help on method raw_input in module ipykernel.kernelbase:

raw_input(prompt='') method of ipykernel.ipkernel.IPythonKernel instance
    Forward raw_input to frontends
    
    Raises
    ------
    StdinNotImplementedError if active frontend doesn't support stdin.



In [22]:
help(sum)

Help on built-in function sum in module builtins:

sum(iterable, /, start=0)
    Return the sum of a 'start' value (default: 0) plus an iterable of numbers
    
    When the iterable is empty, return the start value.
    This function is intended specifically for use with numeric values and may
    reject non-numeric types.



In [23]:
help(str)

Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  

In [24]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



# 6. Function Parameter Types

In Python, function parameters can have different types, and you can specify the type of a parameter using type hints. Type hints are a way to indicate the expected types of function arguments and return values. While Python is a dynamically typed language, type hints provide documentation and can be used by tools like linters and type checkers to catch type-related errors.

Here's how you can specify different types of function parameters using type hints:

1. **Positional Parameters:**

   Positional parameters are the most common type of parameters in Python functions. You can specify their types using the `:` syntax in the function definition.

   ```python
   def add(x: int, y: int) -> int:
       return x + y
   ```

   In this example, `x` and `y` are positional parameters with type hints. They are expected to be integers, and the function is expected to return an integer.

2. **Default Parameters:**

   Default parameters have default values assigned to them, and you can also specify their types.

   ```python
   def greet(name: str, greeting: str = "Hello") -> str:
       return f"{greeting}, {name}!"
   ```

   In this case, `name` is a positional parameter with the type hint `str`, and `greeting` is a default parameter with a default value of `"Hello"` and the type hint `str`.

3. **Keyword Parameters:**

   Keyword parameters are also known as named parameters. They are used when you want to specify the argument's name explicitly when calling the function.

   ```python
   def divide(dividend: float, divisor: float) -> float:
       return dividend / divisor
   ```

   Here, `dividend` and `divisor` are keyword parameters with type hints.

4. **Variable-Length Positional Parameters:**

   You can use the `*args` syntax to create functions that accept a variable number of positional arguments.

   ```python
   def sum_numbers(*args: int) -> int:
       return sum(args)
   ```

   In this case, `*args` is a special syntax that allows you to pass any number of integer arguments.

5. **Variable-Length Keyword Parameters:**

   You can use the `**kwargs` syntax to create functions that accept a variable number of keyword arguments.

   ```python
   def print_info(**kwargs: str) -> None:
       for key, value in kwargs.items():
           print(f"{key}: {value}")
   ```

   Here, `**kwargs` allows you to pass any number of keyword arguments with string keys and string values.

6. **Type Annotations for Variables:**

   You can also use type hints for variables within a function:

   ```python
   def double_and_sum(x: int, y: int) -> int:
       result: int = x * 2 + y
       return result
   ```

   In this example, the `result` variable has a type hint of `int`.

7. **Type Hints for Return Values:**

   You can specify the expected return type of a function using the `->` syntax, as shown in previous examples.

Remember that type hints are not enforced at runtime, but they provide valuable information for code readability, documentation, and static analysis tools like MyPy. Python's "duck typing" philosophy still allows you to pass values of different types to functions, but it's good practice to follow type hints whenever possible to improve code quality and maintainability.

# Practice

In [33]:
# Variable-Length Positional Parameters

def sumofnumbers(*data):
    
    sum = 0
    for i in data:
        sum = sum +i
    return sum

In [34]:
sumofnumbers(1,2,3,4)

10

In [35]:
sumofnumbers(1,2,3,4,5,6,7,8,9,10) # Give Any Number of Input

55

In [36]:
# Variable-Length Keyword Parameters

def sumofnumbers(**kwargs):
    
    print(kwargs,type(kwargs))

In [38]:
sumofnumbers(a=1,b=2,c=3,d=4)

{'a': 1, 'b': 2, 'c': 3, 'd': 4} <class 'dict'>


In [42]:
# Variable-Length Keyword Parameters

def sumofnumbers(**kwargs):
    
    sum = 0
    for i in kwargs.values():
        sum = sum + i
    return sum

In [43]:
sumofnumbers(a=1,b=2,c=3,d=4)

10

In [44]:
sumofnumbers(a=1,b=2,c=3,d=4,e=5,f=100,g=232323)

232438

In [45]:
sumofnumbers(a='str',b=2,c=3,d=4) # Gives Error

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [46]:
# Default Parameters

def some_function(a = 10, b, c = 30): # Gives Error
    return a+b+c

SyntaxError: non-default argument follows default argument (1590229232.py, line 3)

In [47]:
# Default Parameters

def some_function(a = 10, c = 30, b): # Gives Error
    return a+b+c

SyntaxError: non-default argument follows default argument (3034974467.py, line 3)

In [48]:
# Default Parameters

def some_function(b, a = 10, c = 30): # Correct Method
    return a+b+c

In [49]:
some_function(10)

50

In [50]:
some_function(10,10,10)

30

In [51]:
some_function(90,9,1)

100

In [52]:
some_function(100, c = 1)

111

In [53]:
some_function(100, a = 1)

131

In [54]:
some_function(100, a = 0, c = -100)

0

# 7. Default Parameters vs Default Arguments

In Python, "default parameters" and "default arguments" are often used interchangeably, but they refer to slightly different concepts. Let's clarify these terms:

1. **Default Parameters:**
   Default parameters are used in function definitions. When you define a function, you can assign default values to some or all of its parameters. These default values are used when the function is called without providing a value for the corresponding parameter. Default parameters are primarily used to make a function more flexible and to provide sensible or commonly used values when no specific argument is passed.

   Here's an example:

   ```python
   def greet(name, greeting="Hello"):
       print(f"{greeting}, {name}!")

   greet("Alice")  # Uses the default value for greeting
   greet("Bob", "Hi")  # Overrides the default value for greeting
   ```

   In this example, the `greeting` parameter has a default value of "Hello," but you can override it when calling the function.
   

2. **Default Arguments:**

   Default arguments are used when you define a function or method in a more general sense. They refer to arguments that have default values associated with them, which can be overridden when calling the function or method. Default arguments are not limited to just function parameters; they can be used in methods, class constructors, and other contexts where arguments are involved.

   Here's an example using a class method:

   ```python
   class MyClass:
       def __init__(self, value=42):
           self.value = value

   obj1 = MyClass()  # Uses the default value for 'value'
   obj2 = MyClass(10)  # Overrides the default value for 'value'
   ```

   In this example, the `__init__` method of the `MyClass` class takes a default argument `value`, which is set to 42. You can provide a different value when creating an instance of the class.

In summary, default parameters are a specific concept related to function parameters, whereas default arguments are a broader concept that can apply to function parameters, method arguments, and other similar contexts where values can be provided with defaults. The distinction is somewhat subtle, and in practice, the terms are often used interchangeably.

# 8. Lambda Functions

A lambda function in Python is a small, anonymous function defined using the `lambda` keyword. It is often referred to as a "lambda expression" or "lambda function." Lambda functions are typically used for simple operations where you don't want to define a full named function using the `def` keyword. They are especially useful when you need to pass a function as an argument to another function or use it in a functional programming context like `map`, `filter`, or `reduce`.

The basic syntax of a lambda function is as follows:

```python
lambda arguments: expression
```

Here are some detailed examples of how to use lambda functions:

**Example 1: Simple lambda function**

```python
# Define a lambda function to square a number
square = lambda x: x * x

# Use the lambda function
result = square(5)
print(result)  # Output: 25
```

In this example, we define a lambda function that takes one argument (`x`) and returns its square.

**Example 2: Using lambda with `map`**

```python
# Create a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use lambda with map to square each number
squared_numbers = list(map(lambda x: x * x, numbers))

print(squared_numbers)  # Output: [1, 4, 9, 16, 25]
```

Here, we use `map` along with a lambda function to apply the squaring operation to each element in the `numbers` list.

**Example 3: Using lambda with `filter`**

```python
# Create a list of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Use lambda with filter to get even numbers
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))

print(even_numbers)  # Output: [2, 4, 6, 8, 10]
```

In this example, we use `filter` along with a lambda function to filter out only the even numbers from the `numbers` list.

**Example 4: Using lambda for sorting**

```python
# Create a list of tuples with names and ages
people = [("Alice", 30), ("Bob", 25), ("Charlie", 35), ("David", 28)]

# Sort the list by age using lambda
sorted_people = sorted(people, key=lambda x: x[1])

print(sorted_people)
# Output: [('Bob', 25), ('David', 28), ('Alice', 30), ('Charlie', 35)]
```

Here, we use a lambda function as the `key` argument in the `sorted` function to sort the list of people by their ages.

Lambda functions are concise and handy for simple operations, but they should be used judiciously. For more complex operations or functions that need to be reused, it's better to define a regular named function using `def`.

# Practice

In [55]:
# Define a lambda function to multiply a number by 100
multiplyby100 = lambda x: x * 100

# Use the lambda function
result = multiplyby100(5)
print(result)  # Output: 50

500


In [56]:
# Define a lambda function to reverse a string
reverse_string = lambda s: s[::-1]

result = reverse_string("hello")
print(result)  # Output: "olleh"

olleh


In [57]:
# Define a lambda function to check if a number is even or odd
is_even = lambda x: "Even" if x % 2 == 0 else "Odd"

print(is_even(6))  # Output: "Even"
print(is_even(7))  # Output: "Odd"

Even
Odd


In [58]:
# Define a lambda function to calculate the area of a rectangle
rectangle_area = lambda length, width: length * width

area = rectangle_area(5, 4)
print(area)  # Output: 20

20


In [59]:
# Use lambda within a list comprehension to get the squares of numbers
numbers = [1, 2, 3, 4, 5]
squared_numbers = [(lambda x: x * x)(num) for num in numbers]

print(squared_numbers)  # Output: [1, 4, 9, 16, 25]

[1, 4, 9, 16, 25]


In [60]:
# Sort a dictionary by values using lambda
scores = {"Alice": 90, "Bob": 80, "Charlie": 95, "David": 75}
sorted_scores = dict(sorted(scores.items(), key=lambda item: item[1], reverse=True))

print(sorted_scores)
# Output: {'Charlie': 95, 'Alice': 90, 'Bob': 80, 'David': 75}

{'Charlie': 95, 'Alice': 90, 'Bob': 80, 'David': 75}


In this example, we use a lambda function as the key argument in the sorted function to sort a dictionary by its values in descending order.

These examples illustrate how lambda functions can be used for a variety of tasks, from simple calculations to more complex operations, and even in conjunction with built-in Python functions like sorted, map, and filter.

In [61]:
'''You can create a lambda function in Python to sum the odd numbers in a given list
using the filter() and sum() functions. Here's how you can do it'''

# Sample input list
input_list = [1, 2, 3, 4, 5, 6, 7, 8, 9]

# Lambda function to sum odd numbers
sum_odd = lambda lst: sum(filter(lambda x: x % 2 != 0, lst))

# Call the lambda function
result = sum_odd(input_list)

# Print the result
print(result)

25


In this code:

- filter(lambda x: x % 2 != 0, lst) filters the odd numbers from the input list.
- sum() is used to calculate the sum of the filtered odd numbers.

You can replace input_list with your own list when you want to find the sum of odd numbers in a different list.

In [69]:
l = [2,4,3,5]

def sum_odd(l):
    l1 = []
    for i in l:
        if i%2!=0:
            l1.append(i)
    return sum(l1)    

In [70]:
sum_odd(l)

8

In [71]:
[i for i in l if i%2!=0]

[3, 5]

In [72]:
sum([i for i in l if i%2!=0])

8

In [73]:
sum_of_lambda = lambda l: sum([i for i in l if i%2!=0])
sum_of_lambda([2,4,6,3,5,3,5])

16

In [80]:
# Define a lambda function to calculate factorial

factorial = lambda n: 1 if n == 0 else n * factorial(n - 1)

# Input the number for which you want to calculate the factorial

num = int(input("Enter a number: "))

# Calculate and print the factorial

if num < 0:
    print("Factorial is not defined for negative numbers.")
elif num == 0:
    print("The factorial of 0 is 1")
else:
    print(f"The factorial of {num} is {factorial(num)}")

Enter a number: 5
The factorial of 5 is 120


In [81]:
# Factorial Without Lambda Function

def factorial(number):
    
    if number < 0:
        print('Invalid input.')
    elif number == 0:
        return 1
    else:
        return number * factorial(number-1)
    
factorial(5)

120

In [82]:
factorial(-2)

Invalid input.


In [83]:
factorial(1)

1

In [84]:
factorial(4)

24

In [90]:
# Factorial Without Recursion

number, factorial = 5, 1

for index in range(1,number+1):
    factorial = factorial * index
    
print(f'Factorial of {number} is: {factorial}.')

Factorial of 5 is: 120.


In [119]:
# Get Letters in Lower Case Using Filter Function

s = 'PW Skills'
    
list(filter(lambda a: a.islower(), s))

['k', 'i', 'l', 'l', 's']

In [124]:
# Get List Elements with Length Less Than Four

l = ['pw', 'pwskills', 'sdh', 'kris']

list(filter(lambda a: len(a)<=4, l))

['pw', 'sdh', 'kris']

In [128]:
# Get List Elements starting with P or p

l = ['pw', 'pwskills', 'sdh', 'kris', 'PWSKILLS']

list(filter(lambda a: a.startswith('p') or a.startswith('P'), l))

['pw', 'pwskills', 'PWSKILLS']

# 9. Recursive Functions

A recursive function in Python is a function that calls itself in order to solve a problem. This approach is often used for solving problems that can be broken down into smaller, similar subproblems. Recursive functions consist of two main parts: the base case(s) and the recursive case.

Here's how a recursive function works in Python:

**1. Base Case(s):**
   - The base case(s) are the conditions under which the function stops calling itself and returns a result. These conditions are necessary to prevent infinite recursion and ensure that the function terminates.
   - When the function reaches a base case, it returns a result without making any further recursive calls.

**2. Recursive Case:**
   - In the recursive case, the function calls itself with modified arguments to break down the problem into smaller, more manageable subproblems.
   - Each recursive call should move closer to one or more base cases, ensuring that the function eventually reaches the base case(s).

Here's a simple example of a recursive function in Python that calculates the factorial of a non-negative integer:

```python
def factorial(n):
    # Base case: If n is 0 or 1, return 1
    if n == 0 or n == 1:
        return 1
    # Recursive case: n! = n * (n-1)!
    else:
        return n * factorial(n - 1)
```

In this example:

- The base case is when `n` is 0 or 1, where the function returns 1.
- In the recursive case, the function calculates `n!` by multiplying `n` with `factorial(n - 1)`. This is a recursive call that computes the factorial of `(n-1)`, and so on, until it reaches the base case.

When you call `factorial(5)`, for example, it works like this:

1. `factorial(5)` calls `factorial(4)`
2. `factorial(4)` calls `factorial(3)`
3. `factorial(3)` calls `factorial(2)`
4. `factorial(2)` calls `factorial(1)`
5. `factorial(1)` returns 1 (base case)
6. `factorial(2)` returns `2 * 1 = 2`
7. `factorial(3)` returns `3 * 2 = 6`
8. `factorial(4)` returns `4 * 6 = 24`
9. `factorial(5)` returns `5 * 24 = 120`

The key to writing a successful recursive function is to ensure that it converges to the base case(s) and that it breaks down the problem into smaller, more manageable subproblems with each recursive call. Failure to do so may result in infinite recursion and a "RecursionError" in Python.

# Practice

You can calculate Fibonacci numbers using a recursive function in Python. The Fibonacci sequence starts with 0 and 1, and each subsequent number in the sequence is the sum of the two preceding ones. Here's a Python function that calculates Fibonacci numbers recursively:

In [131]:
def fibonacci(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)

# Example usage:
n = 5  # Change this to the desired Fibonacci number you want to calculate
result = fibonacci(n)
print(f"The {n}th Fibonacci number is: {result}")

The 5th Fibonacci number is: 5


This fibonacci function takes an integer n as input and returns the n-th Fibonacci number. Note that this implementation uses recursion, which can be inefficient for large values of n due to repeated calculations. You can optimize it using techniques like memoization or using an iterative approach if you need to calculate Fibonacci numbers for large values of n.

In [133]:
# Sum Till Some Number

def sum_till_n (n):
    if n==1:
        return 1
    else:
        return n+ sum_till_n(n-1)
sum_till_n(5)

15

# 10. Map, Reduce and Filter Functions

In Python, you can use the `map`, `reduce`, and `filter` functions to perform common operations on lists or other iterable data structures. These functions are part of Python's functional programming capabilities and can be quite powerful when used correctly. Below, I'll explain each of these functions and provide examples:

1. `map`:
   - The `map` function applies a given function to each element in an iterable (e.g., a list) and returns an iterable containing the results.

   Syntax:
   ```python
   map(function, iterable)
   ```

   Example:
   ```python
   numbers = [1, 2, 3, 4, 5]

   # Double each element in the list
   doubled_numbers = list(map(lambda x: x * 2, numbers))
   print(doubled_numbers)  # Output: [2, 4, 6, 8, 10]
   ```

2. `filter`:
   - The `filter` function filters elements from an iterable based on a provided function's condition and returns an iterable containing the filtered elements.

   Syntax:
   ```python
   filter(function, iterable)
   ```

   Example:
   ```python
   numbers = [1, 2, 3, 4, 5]

   # Filter even numbers from the list
   even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
   print(even_numbers)  # Output: [2, 4]
   ```

3. `reduce` (requires `functools` module in Python 3):
   - The `reduce` function applies a given function cumulatively to the elements of an iterable, reducing it to a single value. This function is not available directly in Python 3 and requires importing from the `functools` module.

   Syntax:
   ```python
   from functools import reduce

   reduce(function, iterable[, initial])
   ```

   Example:
   ```python
   from functools import reduce

   numbers = [1, 2, 3, 4, 5]

   # Calculate the sum of all elements in the list
   sum_of_numbers = reduce(lambda x, y: x + y, numbers)
   print(sum_of_numbers)  # Output: 15
   ```

Note: In Python 3, you need to convert the result of `map` and `filter` to a list explicitly if you want to see the result as a list, as shown in the examples. Python 2 behaves slightly differently, as `map` and `filter` return a list by default in Python 2.

Keep in mind that in modern Python development, list comprehensions and generator expressions are often preferred over `map`, `filter`, and `reduce` for their readability and performance benefits.

# Practice

In [94]:
numbers = [100, 2, 3, 4, 5]

# Add 2 to each element in the list

addtwo_numbers = list(map(lambda x: x + 2, numbers))
print(addtwo_numbers)

[102, 4, 5, 6, 7]


In [95]:
numbers = [1, 2, 3, 4, 5]

# Filter odd numbers from the list

odd_numbers = list(filter(lambda x: x % 2 != 0, numbers))
print(odd_numbers)

[1, 3, 5]


In [96]:
from functools import reduce

numbers = [1, 2, 3, 4, 5]

# Calculate the product of all elements in the list

product_of_numbers = reduce(lambda x, y: x * y, numbers)
print(product_of_numbers)

120


In [102]:
l = ['sudh', 'kumar', 'skills']

# Use the map function to reverse each element

reversed_list = list(map(lambda x: x[::-1], l))

# Print the reversed list

print(reversed_list)

['hdus', 'ramuk', 'slliks']


In [ ]:
from functools import reduce

numbers = [1, 2, 3, 4, 5]

# Calculate the product of all elements in the list

product_of_numbers = reduce(lambda x, y: x * y, numbers)
print(product_of_numbers)

In [103]:
my_list = [4, 2, 9, 7, 1, 12, 6]

from functools import reduce

def find_min(lst):
    if len(lst) == 0:
        return None  # Handle empty list case

    return reduce(lambda x, y: x if x < y else y, lst)

find_min(my_list)

1

In [105]:
from functools import reduce

def find_max(lst):
    if len(lst) == 0:
        return None  # Handle empty list case

    return reduce(lambda x, y: x if x > y else y, lst)

find_max(my_list)

12

In [109]:
# Factorial Using Reduce Function

from functools import reduce

# Define a function to calculate the factorial of a number
def factorial(n):
    if n == 0:
        return 1
    else:
        # Use reduce to multiply all numbers from 1 to n
        return reduce(lambda x, y: x * y, range(1, n + 1))

# Input the number for which you want to calculate the factorial
num = int(input("Enter a number: "))

# Calculate and print the factorial
result = factorial(num)
print(f"The factorial of {num} is {result}")

Enter a number: 6
The factorial of 6 is 720


In [115]:
numbers = [2,3,4,5,6]

def multiply_two_numbers(x, y):
    return x * y

# Use reduce to find the product of even numbers in a list
product = reduce(multiply_two_numbers, filter(lambda x: x % 2 == 0, numbers))
product

48

In [117]:
reduce(lambda a,b: a*b , [i for i in numbers if i%2==0])

48

# 11. Map Function & For Loop

A `map` function and a `for` loop are both used for iterating over elements in a collection, but they serve slightly different purposes and have distinct characteristics:

1. **Purpose:**

   - **`map`**: The `map` function is typically used to apply a given function to each element of a collection (e.g., a list or an array) and create a new collection with the results. It's used for transforming the elements of a collection without changing the original collection.

   - **`for` loop**: A `for` loop is a general-purpose looping construct that allows you to iterate over a collection of elements or execute a block of code a specific number of times. It can be used for various tasks beyond just transforming elements; you can use it for filtering, summing, searching, and more.

2. **Immutability:**

   - **`map`**: It is generally used in functional programming and often produces a new collection with the transformed elements while leaving the original collection unchanged. This is because it emphasizes immutability, where the original data remains unaltered.

   - **`for` loop**: A `for` loop can be used to modify elements in place within the original collection, which means it can change the original data.

3. **Syntax:**

   - **`map`**: It is a higher-order function that takes a function and a collection as arguments. The function specifies how to transform each element in the collection. The result is a new collection.

   - **`for` loop**: It typically involves writing explicit looping constructs, including initialization, condition, and increment statements. The logic inside the loop determines what happens with each element.

Here's a simple example in Python to illustrate the difference:

Using `map`:
```python
original_list = [1, 2, 3, 4]
transformed_list = list(map(lambda x: x * 2, original_list))
# transformed_list contains [2, 4, 6, 8], and original_list remains [1, 2, 3, 4]
```

Using a `for` loop:
```python
original_list = [1, 2, 3, 4]
for i in range(len(original_list)):
    original_list[i] = original_list[i] * 2
# original_list is now [2, 4, 6, 8]
```

In summary, while both `map` and `for` loops are used for iterating over collections, `map` is more specialized for transforming elements in an immutable way, whereas a `for` loop is a general-purpose construct for various types of iterations and operations. The choice between them depends on the specific task and programming paradigm you're using.